### PR TITLE
chore: optimize zpopminmax operations

### PR DESCRIPTION
**Pull request by romange**
Before, we run FromRank operation count times. This operation computes subtree counts on the fly. Now, we iterate over a tree first, delete the keys from the map - no rank calls. And only after we filled the result, we delete the entries from the tree. For cases where we delete all the entries we avoid calling FromRank.

Moreover, we optimized FromRank operations to handle corner cases of most right and most left nodes by avoiding computing the subtree counts.

### CODE REVIEW

File: src/core/bptree_set.h [hunk 1]
```diff
@@ -435,7 +437,24 @@ void BPTree<T, Policy>::ToRank(uint32_t rank, BPTreePath* path) const {
   assert(root_ && rank < count_);
   BPTreeNode* node = root_;
 
+  if (rank + 1 == count_) {
+    // Corner case where we search for the node on the right.
+    while (!node->IsLeaf()) {
+      path->Push(node, node->NumItems());
+      node = node->Child(node->NumItems());
+    }
+    path->Push(node, node->NumItems() - 1);
```
Comment (dranikpg) [inline src/core/bptree_set.h:446]: with `node->IsLeaf()` that's just 0?

**Review by chakaz — APPROVED** [review]

File: src/core/sorted_map.cc [hunk 2]
```diff
@@ -573,34 +573,50 @@ size_t SortedMap::DeleteRangeByLex(const zlexrangespec& range) {
...
+    count = sz;
 
   res.reserve(count);
+
+  auto cb = [&](ScoreSds obj) {
+    res.emplace_back(string{(sds)obj, sdslen((sds)obj)}, GetObjScore(obj));
+
+    // We can not delete from score_tree because we are in the middle of the iteration.
+    CHECK(score_map->Erase((sds)obj));
+    return true;
```
Comment (dranikpg) [inline src/core/sorted_map.cc:597]: so `return false` was just saved for a future pr?

**Review by dranikpg — DISMISSED** [review]

File: src/core/bptree_set.h (hunk 1 above)
Comment (romange) [reply src/core/bptree_set.h:446]: nope. Consider a root with 2 items. in that case we won't enter the while loop and want to fetch the second one, index=1

File: src/core/sorted_map.cc (hunk 2 above)
Comment (romange) [reply src/core/sorted_map.cc:597]: what do you mean?

**Review by dranikpg — APPROVED** [review]

**Pull request merged**